Sprint 7 solution in file manager

polyakovanan opened the pull request:
Добавил реализацию файлового менеджера задач + покрытие нового функционала тестами.

polyakovanan committed:
Нулевое приближение для решения задачи таск менеджера с файлом. Набросан весь код функционала, но не проверен

polyakovanan committed:
Наработки для тестирования нового функционала

polyakovanan committed:
Доработки нового функционала файлового менеджера задач до рабочего состояния + дополнительное покрытие тестами

polyakovanan committed:
Небольшие корректировки тестов TaskUtils

polyakovanan committed:
Правки по спринту 7.
Корректировка код стайла для автотеста

polyakovanan committed:
Правки по спринту 7.
Корректировка код стайла для автотеста (2)

TyutterinYakov commented on test/ru/terralink/kanban/service/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,415 @@
+import java.nio.file.Paths;
+import java.util.List;
+
+public class FileBackedTaskManagerTest {
```
Также хорошо было бы для классов тестирования менеджеров - ввести общего абстрактного родителя, в котором будут находиться все тесты для `InMemory-реализации`
А уже доп тесты с инициализацией поля `TaskManager` из абстрактного класса - будет в наследниках)
Таким образом мы сможем проверить, что новая надстройка в виде файлового менеджера не сломала стандартное поведение базовой реализации)

polyakovanan replied:
Добавил абстрактный базовый класс TaskManagerTest со всеми тестами для InMemory реализации. Наследовал его в InMemoryTaskManagerTest без каких-либо переопределений и в FileBackedTaskManagerTest, изменив получение самого менеджера в BeforeEach и добавив ссылки на проверочные файлы в переопределении каждого другого метода.

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+        TaskType type;
+        try {
+            type = TaskType.valueOf(elements[1]);
+        } catch (NullPointerException | IllegalArgumentException e) {
```
А из-за чего может возникнуть NPE? Разве при текущей реализации это возможно?)

polyakovanan replied:
Не, не будет такого. Убрал отлов NPE

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+        TaskStatus status;
+        try {
+            status = TaskStatus.valueOf(elements[3]);
+        } catch (NullPointerException | IllegalArgumentException e) {
```
А из-за чего может возникнуть NPE? Разве при текущей реализации это возможно?)

polyakovanan replied:
Убрал отлов NPE

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+        if (type == TaskType.SUBTASK) {
+            try {
+                epicId = Integer.parseInt(elements[5]);
+            } catch (NullPointerException | NumberFormatException e) {
```
А из-за чего может возникнуть NPE? Разве при текущей реализации это возможно?)

polyakovanan replied:
Убрал отлов NPE

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+
+        TaskStatus status;
+        try {
+            status = TaskStatus.valueOf(elements[3]);
```
Можно было бы в enum `TaskStatus` ввести отдельный метод, в котором бы мы прошлись по значениям - `.values()` и, если не нашли совпадений - вернули пустой `Optional`.
Таким образом мы не будем отлавливать исключения, что является не совсем дешевой операцией

polyakovanan replied:
Добавил заглушечный элемент OPTIONAL и метод для обхода values(). Если вернется статус OPTIONAL - считаем данные невалидными

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+
+        TaskType type;
+        try {
+            type = TaskType.valueOf(elements[1]);
```
Можно было бы в enum `TaskType` ввести отдельный метод, в котором бы мы прошлись по значениям - `.values()` и, если не нашли совпадений - вернули пустой `Optional`.
Таким образом мы не будем отлавливать исключения, что является не совсем дешевой операцией

polyakovanan replied:
Добавил заглушечный элемент OPTIONAL и метод для обхода values(). Если вернется тип OPTIONAL - считаем данные невалидными

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+                parsedObject = new Subtask(id, name, description, epicId);
+                parsedObject.setStatus(status);
+            }
+            case EPIC -> parsedObject = new Epic(id, name, description);
```
Не забываем установить тот статус, что был у эпика в файле)
Таким образом потом не нужно будет его пересчитывать для каждой подзадачи в отдельности

polyakovanan replied:
Долго сомневался, что это уместно, поскольку статус эпика считается расчетным значением по постановке из предыдущих спринтов. Но, вроде, мы уже защитили данные от манипуляции извне за счет клонирования, поэтому нарушить целостность данных через setStatus больше не получится. Поэтому сделал это допустимой операцией, используемой только при чтении из файла

TyutterinYakov commented on src/ru/terralink/kanban/service/Managers.java:
```diff
@@ -13,4 +25,36 @@ public static TaskManager getDefault() {
+                } else {
+                    try {
+                        Task task = TaskUtils.fromString(lines[i]);
+                        fileBackedTaskManager.createTask(task);
```
Добавление следует производить напрямую в мапу, чтобы сохранить значение идентификатора, что было в файле, а не переопределить его в методе создания

polyakovanan replied:
Добавил новый метод addParsedTask, который производит вставку напрямую в структуру as-is

TyutterinYakov commented on src/ru/terralink/kanban/util/TaskUtils.java:
```diff
@@ -0,0 +1,68 @@
+
+import ru.terralink.kanban.model.*;
+
+public class TaskUtils {
```
Так как класс не имеет состояния(все `static`), значит создание объектов этого класса не имеет смысла. Можно объявить в нем приватный конструктор без параметров, чтобы избежать создания бессмысленных объектов этого класса. Это тонкость, которую хорошо было бы понять) То есть - если у класса нет публичного конструктора - создание объекта этого класса невозможно(ну почти, в java есть некоторый механизм, который может помочь это обойти, но об этом позже)) Когда ты создаешь класс, в нем по дефолту есть публичный конструктор без параметров, который в коде не видно, но он есть. Если мы создаем приватный, мы его перекрываем, а значит запрещаем создавать объекты этого класса

polyakovanan replied:
Добавил приватный конструктор

TyutterinYakov reviewed (commented):
Добрый вечер! В целом - хорошо) Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего вечера! Если возникнут вопросы по ревью - пиши в пачку, буду рад помочь🌃

polyakovanan committed:
Правки по замечаниям Спринт 7. Итерация 1.
Создал базовый класс для юнит-тестов менеджеров.
Разрешил менять статус Эпика напрямую для целей точного чтения загружаемого файла (все равно изменить статус эпика извне не получится из-за клонирования)
Сделал в TaskUtils приватный конструктор
Реализовал загрузку задач из файла с прямой записью в структуру, без использования методов создания.
Добавил заглушечный тип и статус задачи, чтобы не отлавливать исключение в парсинге

polyakovanan committed:
Правки по замечаниям Спринт 7. Итерация 1.
Исправил один из тестов. Почему-то на автотестах яндекса падает, скорее всего из-за там проверка регистрочувствительная. Ща проверим

TyutterinYakov commented on src/ru/terralink/kanban/service/Managers.java:
```diff
@@ -13,4 +25,37 @@ public static TaskManager getDefault() {
+                } else {
+                    try {
+                        Task task = TaskUtils.fromString(lines[i]);
+                        fileBackedTaskManager.addParsedTask(task);
```
Так же не забываем, что у нас в классе `InMemoryTaskManager` есть такое поле, как `idCounter`, которое хранит наибольший идентификатор. Дак вот при выгрузке он становится 0, хотя он должен иметь значение большего идентификатора среди задач, иначе в мапах могут произойти непредвиденные изменения при создании новых задач/подзадач/эпиков

TyutterinYakov commented on src/ru/terralink/kanban/service/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,116 @@
+
+    public void addParsedTask(Task task) {
+        taskStorage.get(task.getType()).put(task.getId(), task);
+        if (task.getType() == TaskType.SUBTASK) {
```
Здесь можно было бы добавить условие для нахождения большего значения идентификатора)
Например:
```
if (task.getId() > super.idCounter) {
  super.idCounter = task.getId();
}
```

TyutterinYakov reviewed (approved):
Добрый вечер! Поздравляю с успешной сдачей спринта :) Сил и терпения на этом сложном, но интересном пути! Успехов в обучении и прекрасных выходных!🌃🎇

polyakovanan committed:
Правки по замечаниям Спринт 7. Итерация 2.
Добавил обновление idCounter при чтении сохраненных задач из файла.
